base/loading.jl

```diff
@@ -3115,16 +3115,14 @@ end
 
         # now check if this file is fresh relative to its source files
         if !skip_timecheck
-            if !samefile(includes[1].filename, modpath)
-                stdlib_path = fixup_stdlib_path(includes[1].filename)
-                # In certain cases the path rewritten by `fixup_stdlib_path` may
-                # point to an unreadable directory, make sure we can `stat` the
-                # file before comparing it with `modpath`.
-                isreadable = iszero(@ccall jl_fs_access(stdlib_path::Cstring, 0x04::Cint)::Cint)
-                if isreadable && !samefile(stdlib_path, modpath)
-                    @debug "Rejecting cache file $cachefile because it is for file $(includes[1].filename) not file $modpath"
-                    return true # cache file was compiled from a different path
-                end
+            stdlib_path = fixup_stdlib_path(includes[1].filename)
+            # In certain cases the path rewritten by `fixup_stdlib_path` may
+            # point to an unreadable directory, make sure we can `stat` the
+            # file before comparing it with `modpath`.
+            isreadable = iszero(@ccall jl_fs_access(stdlib_path::Cstring, 0x04::Cint)::Cint)
```

Comment:
The only issue is that we now do more work, instead of "just" the happy path of `samefile(filename, modpath) == true`

Reply:
Ok, I brought back the "fast path" that was there before the discussion above, but with what [I understood](https://github.com/JuliaLang/julia/pull/55980#discussion_r1787708185) being the right logic, let me know if I still got it wrong 😅